### DISCUSSION TITLE
Scene-item private_settings?

### DISCUSSION THREAD

**Discussion by MarioMey** [upgradeQ/Streaming-Software-Scripting-Reference#12] [Q&A]
This cheatsheet has [Read and write private data from scripts or plugins](https://github.com/upgradeQ/OBS-Studio-Python-Scripting-Cheatsheet-obspython-Examples-of-API#read-and-write-private-data-from-scripts-or-plugins) part. I didn't tried it, but by reading the code, it writes to `private_settings`... of what? Of OBS itself?

Every source and item has a `"private_settings": {},`as item in JSON. Can't we put data inside it as value of that item?

**Comment by upgradeQ** [comment]
I suggest you to try it, it's basically a shared memory interface (global namespace) for LuaJIT, Python, and C/C++ plugins .
There is a [link](https://github.com/upgradeQ/OBS-Studio-Python-Scripting-Cheatsheet-obspython-Examples-of-API/blob/master/src/export.md) which lists all methods available in `obspython` namespace, you could `Ctrl+F` a word "private" to see how it's used internally in OBS Studio, e.g [this](https://github.com/obsproject/obs-studio/blob/38ad3ba18fc27846e122bd56f589ccb34c4578e2/UI/source-tree.cpp#L507-L512) is how OBS knows that it should collapse or not a group, which reflects  in `scene_collection_name.json` has this section `"private_settings": {
                            "collapsed": true},`

**Reply by MarioMey** [reply]
Oh, there is a `obs_sceneitem_get_private_settings`! I had looked for it and I didn't found it.

But... where is `obs_sceneitem_set_private_settings`? 🤔

**Reply by MarioMey** [reply]
I found where group is collapsed and it would save that state ([here](https://github.com/obsproject/obs-studio/blob/38ad3ba18fc27846e122bd56f589ccb34c4578e2/UI/source-tree.cpp#L523-L534))... but, as I don't understand C++, I don't understand this code:
```c++
void SourceTreeItem::ExpandClicked(bool checked)
{
	OBSData data = obs_sceneitem_get_private_settings(sceneitem);
	obs_data_release(data);

	obs_data_set_bool(data, "collapsed", checked);

	if (!checked)
		tree->GetStm()->ExpandGroup(sceneitem);
	else
		tree->GetStm()->CollapseGroup(sceneitem);
}
```
Why does it create `data` object... and it realeases it inmediatly? Then, it set `collapsed` setting to that data object... but when does that data is set to sceneitem?

**Reply by upgradeQ** [reply]
I guess it's one-line C++ equivalent to Python context managers, [here](https://github.com/obsproject/obs-studio/blob/2eca4d80b681e417cc8966987c59bfc645726d02/UI/window-basic-main.cpp#L7900-L7905) is more illustrative example without auto release magic.

**Reply by upgradeQ** [reply]
>Then, it set collapsed setting to that data object... but when does that data is set to sceneitem?

I think it's used for initialization and saving state. There is no bindings to change that while runtime, [this is not exposed](https://github.com/obsproject/obs-studio/blob/dde4d57d726ed6d9e244ffbac093d8ef54e29f44/UI/obs-frontend-api/obs-frontend-api.h#L94-L95) to python also.
